rust/agama-cli/src/logs.rs

```diff
@@ -0,0 +1,257 @@
+        match copy_items(
+            &[self.src_path.as_str()],
+            self.to().parent().unwrap(),
+            &options,
+        ) {
+            Ok(_p) => Ok(()),
+            Err(_e) => Err(io::Error::new(
+                io::ErrorKind::Other,
+                "Copying of a file failed",
+            )),
+        }
+    }
+}
+
+impl LogItem for LogCmd {
+    fn from(&self) -> &String {
+        &self.cmd
+    }
+
+    fn to(&self) -> PathBuf {
+        self.dst_path.as_path().join(format!("{}", self.cmd))
+    }
+
+    fn store(&self) -> Result<(), Error> {
+        let cmd_parts = self.cmd.split_whitespace().collect::<Vec<&str>>();
+        let file_path = self.to();
+        let output = Command::new(cmd_parts[0])
+            .args(cmd_parts[1..].iter())
+            .output()
+            .expect("Failed to run the command");
+        let mut file_stdout = File::create(format!("{}.out.log", file_path.display()))?;
+        let mut file_stderr = File::create(format!("{}.err.log", file_path.display()))?;
+
+        file_stdout.write_all(&output.stdout)?;
+        file_stderr.write_all(&output.stderr)?;
+
+        Ok(())
+    }
+}
+
+fn store(verbose: bool) -> Result<(), io::Error> {
+    if !Uid::effective().is_root() {
+        panic!("No Root, no logs. Sorry.");
+    }
+
+    // 0. preparation, e.g. in later features some log commands can be added / excluded per users request or
+    let commands = DEFAULT_COMMANDS;
+    let paths = DEFAULT_PATHS;
+    let result = format!("{}.{}", DEFAULT_RESULT, DEFAULT_COMPRESSION.1);
+    let noisy = verbose;
+    let compression = DEFAULT_COMPRESSION.0;
+    let mut log_sources: Vec<Box<dyn LogItem>> = Vec::new();
+
+    showln(noisy, "Collecting Agama logs:");
+
+    // 1. create temporary directory where to collect all files (similar to what old save_y2logs
+    // does)
+    let tmp_dir = TempDir::new(DEFAULT_TMP_DIR)?;
+
+    // 2. collect existing / requested paths which should already exist
+    showln(noisy, "\t- proceeding well known paths");
+    for path in paths {
+        // assumption: path is full path
+        if Path::new(path).try_exists().is_ok() {
+            log_sources.push(Box::new(LogPath::new(path, tmp_dir.path())));
+        }
+    }
+
+    // 3. some info can be collected via particular commands only
+    showln(noisy, "\t- proceeding output of commands");
+    for cmd in commands {
+        log_sources.push(Box::new(LogCmd::new(cmd, tmp_dir.path())));
+    }
+
+    // 4. store it
+    showln(true, format!("Storing result in: \"{}\"", result).as_str());
+
+    for log in log_sources.iter() {
+        show(
+            noisy,
+            format!("\t- storing: \"{}\" ... ", log.from()).as_str(),
+        );
+
+        // for now keep directory structure close to the original
+        // e.g. what was in /etc will be in /<tmp dir>/etc/
+        let res = match fs::create_dir_all(log.to().parent().unwrap()) {
```

Comment:
What about the following?

```rust
        fs::create_dir_all(log.to().parent().unwrap())?;
        let res = match log.store() {
            Ok(_p) => "[Ok]",
            Err(_e) => "[Failed]",
        };
```

Reply:
I'm still not sure if i get it correctly. Does it mean that if ```?``` operator's result is an Err then it returns immediately? (i don't want to call store when create_dir_all failes)

Reply:
Yes, the `?` operator returns immediately. Additionally, it tries to convert the error type to the one returned by the function.

Reply:
but if it exists whole function, then it doesn't try to store other logs (the create dir is log specific in this case) ... it might but even might not fail for the others ....